File: TypeScriptAndAngularProjects/typescript/solutions/VideoStore/src/store/customer.ts
import {Rental} from './rental'

export class Customer {
    rentals: Rental[];

    constructor(public name: string) {
        this.rentals = [];
    }

    addRental(rental: Rental): void {
        this.rentals.push(rental);
    }

    addRentals(...rentals: Rental[]): void {
        rentals.forEach(x => this.addRental(x));
    }

    totalCost(): number {
        let totalAmount = 0;
        for (let rental of this.rentals) {
            totalAmount += rental.cost;
        }
        return totalAmount;
    }

    totalPoints(): number {
        let allPoints = this.rentals.map(r => r.points);

        //this should be a reduction e.g. using lodash
        let points = 0;
        allPoints.forEach(p => points += p);
        return points;
    }

    statement(): string {
        let result = this.getHeader();
        this.rentals.forEach(x => result += this.getStatementRow(x));
        result += `Amount owed is ${this.totalCost()}\n`;
        result += `You earned ${this.totalPoints()} frequent renter points\n`;

        return result;
    }

    private getStatementRow({daysRented, movie, cost}: Rental): string {
        return `\t${daysRented}\t${movie.title}\t${cost}\n`;
    }

    private getHeader(): string {
        return `\nRental Record for ${this.name}\n`;
    }
}

File: JavaProjects/JavaLanguageExercises/src/videostore/finish/two/Customer.java
/*
 * This example adapted from 'Refactoring' by Martin Fowler
 *
 */
package videostore.finish.two;

import java.util.ArrayList;
import java.util.List;

public class Customer {
    private final String name;
    private final List<Rental> rentals = new ArrayList<Rental>();

    public Customer(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public void addRental(Rental arg) {
        rentals.add(arg);
    }

    private double totalCost() {
        return rentals.stream()
                .mapToDouble(Rental::cost)
                .sum();
    }

    private int totalPoints() {
        return rentals.stream()
                .mapToInt(Rental::points)
                .sum();
    }

    public String statement() {
        StringBuilder result = new StringBuilder();
        addHeaderLine(result);
        rentals.forEach(rental -> addRentalLine(result, rental));
        addFooterLines(result);
        return result.toString();
    }

    private void addFooterLines(StringBuilder result) {
        result.append("Amount owed is ");
        result.append(totalCost());
        result.append("\n");
        result.append("You earned ");
        result.append(totalPoints());
        result.append(" frequent renter points\n");
    }

    private void addRentalLine(StringBuilder result, Rental rental) {
        result.append("\t");
        result.append(rental.getDaysRented());
        result.append("\t");
        result.append(rental.getMovie().getTitle());
        result.append("\t");
        result.append(rental.cost());
        result.append("\n");
    }

    private void addHeaderLine(StringBuilder result) {
        result.append("\nRental Record for ");
        result.append(getName());
        result.append("\n");
    }
}